File: LeetCode/JavaScript/src/three_sum_closest.js
var threeSumClosest = function (nums, target) {
    // Sort the array
    nums.sort((a, b) => a - b);
    // Length of the array
    const n = nums.length;
    // Result
    let closest = nums[0] + nums[1] + nums[n - 1];
    // Loop for each element of the array
    for (let i = 0; i < n - 2; i++) {
        // Left and right pointers
        let j = i + 1;
        let k = n - 1;
        // Loop for all other pairs
        while (j < k) {
            let sum = nums[i] + nums[j] + nums[k];
            if (sum <= target) {
                j++;
            } else {
                k--;
            }
            if (Math.abs(closest - target) > Math.abs(sum - target)) {
                closest = sum;
            }
        }
    }
    return closest;
};


console.log(threeSumClosest([-1, 2, 1, -4], 1));
console.log(threeSumClosest([2, 3, 7, 1, 5, 7, 1, 8, -1, -4, -4, 3], 12))

File: LeetCode/Python/src/Three_Sum_Closest.py
"""
@author - Anirudh Sharma
"""
from typing import List


def threeSumClosest(nums: List[int], target: int) -> int:
    # Sort the given array
    nums.sort()
    # Length of the array
    n = len(nums)
    # Closest value
    closest = nums[0] + nums[1] + nums[n - 1]
    # Loop for each element of the array
    for i in range(0, n - 2):
        # Left and right pointers
        j = i + 1
        k = n - 1
        # Loop for all other pairs
        while j < k:
            current_sum = nums[i] + nums[j] + nums[k]
            if current_sum <= target:
                j += 1
            else:
                k -= 1
            if abs(closest - target) > abs(current_sum - target):
                closest = current_sum
    return closest


if __name__ == '__main__':
    print(threeSumClosest([2, 3, 7, 1, 5, 7, 1, 8, -1, -4, -4, 3], 12))
